lib/puma/cluster/fork_pipe.rb

```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require "io/wait"
+
+module Puma
+  class Cluster
+    class ForkPipeReader
+      PAYLOAD_STRING = "q" # going with bigint out of learned fear of overflow
+      PAYLOAD_SIZE = 8 # # of bytes in a q payload
+
+      RESTART_SERVER = 0
+      START_REFORK = -1
+      AFTER_REFORK = -2
+
+      # avoids allocation of objects while reading
+      # only reads 1 payload at a time (minimize data loss if process exits unexpectedly)
+      # NOT thread-safe, but concurrent access not required anywhere currently
+      def initialize(pipe)
+        @pipe = pipe
+        @buffer = +""
+        @payload = +""
+      end
+
+      def read_nonblock
+        @buffer.clear
+        @payload.clear
+        remaining = PAYLOAD_SIZE
+        while remaining > 0
+          case @pipe.read_nonblock(remaining, @buffer, exception: false)
+          when :wait_readable
+            @pipe.wait_readable
+          when :wait_writable
+            @pipe.wait_writable
+          else
```

Comment:
Considering that this implementation is specific to `IO.pipe`, do we need to concern ourselves with `IO::WaitWritable` here? Reading through the docs it seems it's only necessary when the io may be a `OpenSSL::SSL::SSLSocket`. Same question applies to the writer and `:wait_readable` below.

https://docs.ruby-lang.org/en/3.4/IO.html#method-i-read_nonblock

Reply:
Both _nonblock methods are extraneous for now so I'll probably just remove them; it's true that if these are limited to pipes the `:wait_writable` isn't necessary, that was just force of habit.